gcloud-java-compute/src/main/java/com/google/gcloud/compute/DiskConfiguration.java

```diff
@@ -0,0 +1,203 @@
+    }
+
+    Builder(Type type, DiskConfiguration diskConfiguration) {
+      this.type = type;
+      this.sizeGb = diskConfiguration.sizeGb;
+      this.diskType = diskConfiguration.diskType;
+    }
+
+    Builder(Type type, Disk diskPb) {
+      this.type = type;
+      this.sizeGb = diskPb.getSizeGb();
+      if (diskPb.getType() != null) {
+        this.diskType = DiskTypeId.fromUrl(diskPb.getType());
+      }
+    }
+
+    @SuppressWarnings("unchecked")
+    protected B self() {
+      return (B) this;
+    }
+
+    B type(Type type) {
+      this.type = type;
+      return self();
+    }
+
+    /**
+     * Sets the size of the persistent disk, in GB.
+     */
+    public B sizeGb(Long sizeGb) {
+      this.sizeGb = sizeGb;
+      return self();
+    }
+
+    /**
+     * Sets the identity of the disk diskType.
+     */
+    public B diskType(DiskTypeId diskType) {
+      this.diskType = diskType;
+      return self();
+    }
+
+    /**
+     * Creates an object.
+     */
+    public abstract T build();
+  }
+
+  DiskConfiguration(Builder builder) {
+    this.type = builder.type;
+    this.sizeGb = builder.sizeGb;
+    this.diskType = builder.diskType;
+  }
+
+  /**
+   * Returns the disk configuration's type. This method returns {@link Type#STANDARD} for a standard
+   * configuration that creates a disk given its type and size. This method returns
+   * {@link Type#SNAPSHOT} for a configuration that creates a disk from a Google Compute Engine
+   * snapshot. This method returns {@link Type#IMAGE} for a configuration that creates a disk
+   * from a Google Compute Engine image.
+   */
+  public Type type() {
+    return type;
+  }
+
+  /**
+   * Returns the size of the persistent disk, in GB.
+   */
+  public Long sizeGb() {
+    return sizeGb;
+  }
+
+  /**
+   * Returns the identity of the disk type.
+   */
+  public DiskTypeId diskType() {
+    return diskType;
+  }
+
+  /**
+   * Returns a builder for the object.
+   */
+  public abstract Builder toBuilder();
+
+  ToStringHelper toStringHelper() {
+    return MoreObjects.toStringHelper(this)
+        .add("type", type)
+        .add("sizeGb", sizeGb)
+        .add("diskType", diskType);
+  }
+
+  @Override
+  public String toString() {
+    return toStringHelper().toString();
+  }
+
+  final int baseHashCode() {
+    return Objects.hash(type, sizeGb, diskType);
+  }
+
+  final boolean baseEquals(DiskConfiguration diskConfiguration) {
+    return diskConfiguration != null
+        && getClass().equals(diskConfiguration.getClass())
+        && Objects.equals(toPb(), diskConfiguration.toPb());
+  }
+
+  abstract DiskConfiguration setProjectId(String projectId);
+
+  Disk toPb() {
+    Disk diskPb = new Disk();
+    diskPb.setSizeGb(sizeGb);
+    if (diskType != null) {
+      diskPb.setType(diskType.selfLink());
+    }
+    return diskPb;
+  }
+
+  @SuppressWarnings("unchecked")
+  static <T extends DiskConfiguration> T fromPb(Disk diskPb) {
+    if (diskPb.getSourceImage() != null) {
```

Comment:
Fields selection will come in a future PR (I assume); we plan to always choose the sourceImage, sourceSnapshot, and diskConfiguration fields, correct?  (That way we won't accidentally default to a `StandardDiskConfiguration`)